src/proxy/pool.rs

```diff
@@ -12,241 +12,1413 @@
+        let conn_drop_counter: Arc<AtomicU32> = Arc::new(AtomicU32::new(0));
+        let (server_addr, server_handle) = spawn_server(
+            server_drain,
+            conn_counter.clone(),
+            conn_drop_counter.clone(),
+        )
+        .await;
+
+        let cfg = crate::config::Config {
+            local_node: Some("local-node".to_string()),
+            pool_max_streams_per_conn: 1000,
+            pool_unused_release_timeout: Duration::from_secs(2),
+            ..crate::config::parse_config().unwrap()
+        };
+        let sock_fact = Arc::new(crate::proxy::DefaultSocketFactory);
+        let cert_mgr = identity::mock::new_secret_manager(Duration::from_secs(10));
+        let pool = WorkloadHBONEPool::new(cfg.clone(), sock_fact, cert_mgr);
+
+        let mut key1 = WorkloadKey {
+            src_id: Identity::default(),
+            dst_id: vec![Identity::default()],
+            src: IpAddr::from([127, 0, 0, 1]),
+            dst: server_addr,
+        };
+
+        let client_count = 100;
+        let mut count = 0u32;
+        let mut tasks = futures::stream::FuturesUnordered::new();
+        loop {
+            count += 1;
+            if count % 2 == 0 {
+                debug!("using key 2");
+                key1.src = IpAddr::from([127, 0, 0, 4]);
+            } else if count % 3 == 0 {
+                debug!("using key 3");
+                key1.src = IpAddr::from([127, 0, 0, 6]);
+            } else {
+                debug!("using key 1");
+                key1.src = IpAddr::from([127, 0, 0, 2]);
+            }
+
+            tasks.push(spawn_client(pool.clone(), key1.clone(), server_addr, 100));
+
+            if count == client_count {
+                break;
+            }
+        }
+        while let Some(Err(res)) = tasks.next().await {
+            assert!(!res.is_panic(), "CLIENT PANICKED!");
+            continue;
+        }
+
+        let before_conncount = conn_counter.load(Ordering::Relaxed);
+        let before_dropcount = conn_drop_counter.load(Ordering::Relaxed);
+        assert!(
+            before_conncount == 3,
+            "actual before conncount was {before_conncount}"
+        );
+        assert!(
+            before_dropcount == 0,
+            "actual before dropcount was {before_dropcount}"
+        );
+
+        // Attempt to wait long enough for pool conns to timeout+evict
+        sleep(Duration::from_secs(2)).await;
+
+        let real_conncount = conn_counter.load(Ordering::Relaxed);
+        let real_dropcount = conn_drop_counter.load(Ordering::Relaxed);
+        assert!(real_conncount == 3, "actual conncount was {real_conncount}");
+        assert!(real_dropcount == 3, "actual dropcount was {real_dropcount}");
+
+        server_drain_signal.drain().await;
+        server_handle.await.unwrap();
+        drop(pool);
+    }
+
+    #[tokio::test]
+    async fn test_pool_100_clients_evicts_but_does_not_close_active_conn() {
```

Comment:
This is pretty slow, about 8s on my machine. Seems like we can drop this down a ton. I played around with it a bit:

```
diff --git a/src/proxy/pool.rs b/src/proxy/pool.rs
index c2a4294..cb6f744 100644
--- a/src/proxy/pool.rs
+++ b/src/proxy/pool.rs
@@ -629,6 +629,7 @@ mod test {
     use tokio::net::TcpListener;
     use tokio::task::{self};
     use tokio::time::sleep;
+    use tracing::info;
 
     #[cfg(tokio_unstable)]
     use tracing::Instrument;
@@ -1149,7 +1150,7 @@ mod test {
 
     #[tokio::test]
     async fn test_pool_100_clients_evicts_but_does_not_close_active_conn() {
-        // crate::telemetry::setup_logging();
+        crate::telemetry::setup_logging();
 
         let (server_drain_signal, server_drain) = drain::channel();
 
@@ -1165,7 +1166,7 @@ mod test {
         let cfg = crate::config::Config {
             local_node: Some("local-node".to_string()),
             pool_max_streams_per_conn: 50,
-            pool_unused_release_timeout: Duration::from_secs(2),
+            pool_unused_release_timeout: Duration::from_millis(200),
             ..crate::config::parse_config().unwrap()
         };
         let sock_fact = Arc::new(crate::proxy::DefaultSocketFactory);
@@ -1181,58 +1182,63 @@ mod test {
         let client_count = 100;
         let mut count = 0u32;
         let mut tasks = futures::stream::FuturesUnordered::new();
+        info!("run0");
         loop {
             count += 1;
-            tasks.push(spawn_client(pool.clone(), key1.clone(), server_addr, 100));
+            tasks.push(spawn_client(pool.clone(), key1.clone(), server_addr, 1));
 
             if count == client_count {
                 break;
             }
         }
 
+        info!("run1");
         let (client_stop_signal, client_stop) = drain::channel();
         let persist_res =
             spawn_persistent_client(pool.clone(), key1.clone(), server_addr, client_stop);
+        info!("run12");
 
+        futures::future::try_join_all(tasks).await.unwrap();
         //loop thru the nonpersistent clients and wait for them to finish
-        while let Some(Err(res)) = tasks.next().await {
-            assert!(!res.is_panic(), "CLIENT PANICKED!");
-            continue;
-        }
+        // while let Some(Err(res)) = tasks.next().await {
+        //     assert!(!res.is_panic(), "CLIENT PANICKED!");
+        //     continue;
+        // }
+        info!("run3");
 
         let before_conncount = conn_counter.load(Ordering::Relaxed);
         let before_dropcount = conn_drop_counter.load(Ordering::Relaxed);
         assert!(
-            before_conncount == 4,
+            before_conncount == 2,
             "actual before conncount was {before_conncount}"
         );
         assert!(
-            before_dropcount == 0,
+            before_dropcount == 1,
             "actual before dropcount was {before_dropcount}"
         );
 
         // Attempt to wait long enough for pool conns to timeout+evict
-        sleep(Duration::from_secs(2)).await;
+        sleep(Duration::from_millis(200)).await;
 
         let real_conncount = conn_counter.load(Ordering::Relaxed);
-        assert!(real_conncount == 4, "actual conncount was {real_conncount}");
+        assert!(real_conncount == 3, "actual conncount was {real_conncount}");
         // At this point, we should still have one conn that hasn't been dropped
         // because we haven't ended the persistent client
         let real_dropcount = conn_drop_counter.load(Ordering::Relaxed);
-        assert!(real_dropcount == 3, "actual dropcount was {real_dropcount}");
+        assert!(real_dropcount == 2, "actual dropcount was {real_dropcount}");
         client_stop_signal.drain().await;
         assert!(persist_res.await.is_ok(), "PERSIST CLIENT ERROR");
 
-        sleep(Duration::from_secs(2)).await;
+        sleep(Duration::from_millis(200)).await;
 
         let after_conncount = conn_counter.load(Ordering::Relaxed);
         assert!(
-            after_conncount == 4,
+            after_conncount == 3,
             "after conncount was {after_conncount}"
         );
         let after_dropcount = conn_drop_counter.load(Ordering::Relaxed);
         assert!(
-            after_dropcount == 4,
+            after_dropcount == 3,
             "after dropcount was {after_dropcount}"
         );
         server_drain_signal.drain().await;

```
Gets to 500ms

The concerning bit is the assertions are different... but they seem more correct to me somehow. Probably I don't understand the test fully.

Reply:
Yeah the conn count changes because with streamcount = 100, when each client "takes" the connection and we check the streamcount (and what it happens to be at that time based on currently-sending clients) might vary.

Dropping this down to 1 stream per will make this test more deterministic. Unfortunately, doing so for me causes the test server to start refusing conns, probably because we are making too many connection attempts too quickly.

Reply:
I halved most of these.

> test result: ok. 10 passed; 0 failed; 0 ignored; 0 measured; 125 filtered out; finished in 2.53s

The test rig is crude (especially the fake server) and it could be made faster.